runtime/src/parachains.rs

```diff
@@ -324,36 +247,23 @@ decl_module! {
 				);
 
 				Self::dispatch_upward_messages(
-					current_number,
-					&active_parachains,
 					MAX_QUEUE_COUNT,
 					WATERMARK_QUEUE_SIZE,
 					Self::dispatch_message,
 				);
 			}
 
-			<DidUpdate>::put(true);
+			DidUpdate::put(proceeded);
 
 			Ok(())
 		}
 
-		/// Register a parachain with given code.
-		/// Fails if given ID is already used.
-		#[weight = SimpleDispatchInfo::FixedOperational(5_000_000)]
-		pub fn register_parachain(origin, id: ParaId, code: Vec<u8>, initial_head_data: Vec<u8>) -> Result {
-			ensure_root(origin)?;
-			<Self as ParachainRegistrar<T::AccountId>>::register_parachain(id, code, initial_head_data)
+		fn on_initialize() {
+			<Self as Store>::DidUpdate::kill();
 		}
 
-		/// Deregister a parachain with given id
-		#[weight = SimpleDispatchInfo::FixedOperational(10_000)]
-		pub fn deregister_parachain(origin, id: ParaId) -> Result {
-			ensure_root(origin)?;
-			<Self as ParachainRegistrar<T::AccountId>>::deregister_parachain(id)
-		}
-
-		fn on_finalize(_n: T::BlockNumber) {
-			assert!(<Self as Store>::DidUpdate::take(), "Parachain heads must be updated once in the block");
+		fn on_finalize() {
+			assert!(<Self as Store>::DidUpdate::exists(), "Parachain heads must be updated once in the block");
```

Comment:
can be optimised by replacing with a `::take`, but that would mean refactoring so that `registrar::on_finalize` is always called before `parachains::on_finalize`. i'd prefer to leave such optimisations for later.